### PR TITLE
Open MailKit secure socket option setting

### PR DESCRIPTION
**Pull request by arthurchiakh**
related to Issue #69 

Enable user to set secure socket option for MailKit

### CODE REVIEW

**Comment by arthurchiakh** [comment]
Hi, @nblumhardt ,
do you mind checking this PR? 
This change is to let user to specify secure socket option setting for MailKit, which can increase the range of supported email servers for this library.
In my case, I can only connect to Office365 email server when the secure socket option set to SecureSocketOptions.StartTls.
Please see also related issue #69 **Mailkit client use SecureSocketOptions** .

**Comment by Gondost** [comment]
I would like to see this merged as well, I need to set the SecureSocketOptions in order to send email in my environment.

**Comment by nblumhardt** [comment]
Thanks for the nudge. I would prefer not to expose `MailKit` types directly on the email sink's API - is there a higher-level way this can be expressed, akin to how we currently set `EnableSSL`?

**Comment by Gondost** [comment]
@nblumhardt 

`EnableSSL` is a bit simpler because it's a boolean on both ends, whereas `SecureSocketOptions` is an enum. At some point, the enum will have to be validated to ascertain whether it's valid. This seems to lead to paths like creating your own enum with the same values as `SecureSocketOptions`, etc. Or, casting it in some manner and doing an automatic mapping. One way or the other, enum/type mapping seems unavoidable, so it seems like it's a matter of how you prefer to structure the sink code.

**Comment by andriysavin** [comment]
> Thanks for the nudge. I would prefer not to expose `MailKit` types directly on the email sink's API - is there a higher-level way this can be expressed, akin to how we currently set `EnableSSL`?

I think this sink should introduce an enum similar or the same to SecureSocketOptions to avoid dependency on MailKit. And have one-to-one mapping between these enums. Come on, guys, lets finish this, it will be very useful!

**Comment by nblumhardt** [comment]
Introducing an enum that maps to MailKit's one sounds reasonable 👍

**Comment by arthurchiakh** [comment]
I'm sorry guys, better late than never.
I have added a dedicated Enum to map the MailKit socket options.

@nblumhardt please review my latest change. Thanks loads.

File: src/Serilog.Sinks.Email/Sinks/Email/MailKitSecureSocketOptions.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
+﻿#if MAIL_KIT
+
+namespace Serilog.Sinks.Email.Sinks.Email
```
Comment (andriysavin) [inline src/Serilog.Sinks.Email/Sinks/Email/MailKitSecureSocketOptions.cs:3]: Duplicate in the namespace

File: src/Serilog.Sinks.Email/Sinks/Email/MailKitSecureSocketOptions.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+{
+    public enum MailKitSecureSocketOptions
+    {
+        //
```
Comment (andriysavin) [inline src/Serilog.Sinks.Email/Sinks/Email/MailKitSecureSocketOptions.cs:7]: Could you wrap the docs in xml tags so they are shown by intellisense?

File: src/Serilog.Sinks.Email/Sinks/Email/MailKitSecureSocketOptions.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
+﻿#if MAIL_KIT
+
+namespace Serilog.Sinks.Email.Sinks.Email
+{
```
Comment (andriysavin) [inline src/Serilog.Sinks.Email/Sinks/Email/MailKitSecureSocketOptions.cs:4]: I believe the goal of adding this enum was to hide dependency on MailKit. So maybe it should be named just "SecureSocketOptions"?

**Comment by arthurchiakh** [comment]
@nblumhardt  @andriysavin 
I already changed the code accordingly. 
Please let me know anything that I can improve. 
Thanks.

**Review by andriysavin — APPROVED** [review]

**Comment by arthurchiakh** [comment]
@nblumhardt Hi,
can you have a look on this?

**Comment by andriysavin** [comment]
@nblumhardt please, take a look

File: src/Serilog.Sinks.Email/Sinks/Email/SecureSocketOptions.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
+﻿#if MAIL_KIT
```
Comment (nblumhardt) [inline src/Serilog.Sinks.Email/Sinks/Email/SecureSocketOptions.cs:1]: This file should attribute the original source in MailKit (including its license header or link to the MailKit license).

File: src/Serilog.Sinks.Email/Sinks/Email/SecureSocketOptions.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    /// </remarks>
+    public enum SecureSocketOptions
+    {
+        /// <summary>
+        /// No SSL or TLS encryption should be used.
+        /// </summary>
+        None = 0,
+
+        /// <summary>
+        /// Allow the MailKit.IMailService to decide which SSL or TLS options to use (default).
```
Comment (nblumhardt) [inline src/Serilog.Sinks.Email/Sinks/Email/SecureSocketOptions.cs:18]: > Allow the implementation to decide ...

?

File: src/Serilog.Sinks.Email/Sinks/Email/SecureSocketOptions.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+        /// <summary>
+        /// The connection should use SSL or TLS encryption immediately.
+        /// </summary>
+        SslOnConnect = 2,
+
+        /// <summary>
+        /// Elevates the connection to use TLS encryption immediately after reading the greeting
+        /// and capabilities of the server. If the server does not support the STARTTLS extension,
+        /// then the connection will fail and a System.NotSupportedException will be thrown.
```
Comment (nblumhardt) [inline src/Serilog.Sinks.Email/Sinks/Email/SecureSocketOptions.cs:32]: The exception won't propagate out of Serilog, here, so `and a System.NotSupportedException will be thrown.` can probably just be trimmed off.

**Review by nblumhardt — CHANGES_REQUESTED** [review]
Looking good 👍

**Comment by nblumhardt** [comment]
Righto! 👍 Let's give this a try on `dev` 😎

**Comment by andriysavin** [comment]
@nblumhardt Is there a package feed for dev branch?

**Comment by arthurchiakh** [comment]
> @nblumhardt Is there a package feed for dev branch?

I believe this is the one.
https://www.nuget.org/packages/Serilog.Sinks.Email/2.4.1-dev-00128

**Comment by lenlenya** [comment]
Thank you for your contribution, It's very helpful to me.

**Pull request merged**